zha: Clean up listener registration and button state

rcloran opened the pull request:
- Instead of registering listeners in the entity __init__, do it in
   async_added_to_hass to avoid errors updating an entity which isn't fully
   set up yet
 - schedule_update_ha_state -> async_schedule_update_ha_state

## Description:


**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#<home-assistant.github.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

MartinHjelmare commented on homeassistant/components/binary_sensor/zha.py:
```diff
@@ -227,20 +226,20 @@ def move_level(self, change):
         self._level = min(255, max(0, self._level + change))
         self._state = bool(self._level)
-        self.schedule_update_ha_state()
+        self.async_schedule_update_ha_state()
```
We need to know what is calling the callback, a thread or the event loop. That was my question.

rcloran replied:
The event loop

rcloran replied:
Since that is a coro, doesn’t it need to be scheduled anyways?

rcloran replied:
Doh, I was reading async_update_ha_state, not async_schedule_update_ha_state. Alright, this all makes much more sense now.

MartinHjelmare commented on homeassistant/components/zha/__init__.py:
```diff
@@ -321,19 +316,30 @@ def __init__(self, endpoint, in_clusters, out_clusters, manufacturer,
+        # to a dict of cluster ID -> listener to receive messages for specific
+        # clusters separately
+        self._in_listeners = {}
+        self._out_listeners = {}
```
This will reset the listeners added in the child. You can init the parent before the setting the rest of the child's attributes.

rcloran committed:
zha: Clean up binary_sensor listener registration/state updates

 - Instead of registering listeners in the entity __init__, do it in
   async_added_to_hass to avoid errors updating an entity which isn't fully
   set up yet
 - Change from schedule_update_ha_state to async_schedule_update_ha_state

MartinHjelmare reviewed (approved):
Great!
